lmdeploy/pytorch/kernels/dipu/apply_rotary_pos_emb.py

```diff
@@ -0,0 +1,32 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+import deeplink_ext.cpp_extensions as ext
+import torch
+from torch import Tensor
+
+
+def apply_rotary_pos_emb(
+    query_states: Tensor,
+    key_states: Tensor,
+    cos: Tensor,
+    sin: Tensor,
+    position_ids: Tensor,
+    position_ids_1d: Tensor,
+    q_embed=None,
+    k_embed=None,
+    context=None,
+):
+    bs, head, dim = query_states.shape
+    numKeyValueHeads = key_states.shape[1]
+    seqlens = [(min(position_id), max(position_id) + 1)
```

Comment:
can we put seqlens in `if not (hasattr(context, 'cos'))....` branch?

Reply:
I will index cos and sin with position_ids_1d in the next commit.